File: LeetCode/remote/sort_list.cs
//  https://leetcode.com/problems/sort-list/
//
//  Sort a linked list in O(n log n) time using constant space complexity.

using System;

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
    
    public override String ToString()
    {
        return String.Format("{0} {1}", val, next == null ? String.Empty : next.ToString());
    }
}

public class Solution {
    public ListNode SortList(ListNode head) {
        if (head == null || head.next == null)
        {
            return head;
        }

        var slow = head;
        var fast = head.next.next;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        var right = SortList(slow.next);
        slow.next = null;
        var left = SortList(head);
        return Merge(left, right);
    }

    public ListNode Merge(ListNode a, ListNode b)
    {
        if (a == null || b == null)
        {
            return b == null ? a : b;
        }
    
        if (a.val < b.val)
        {
            a.next = Merge(a.next, b);
            return a;
        }
        else
        {
            b.next = Merge(a, b.next);
            return b;
        }
    }

    static void Main()
    {
        var s = new Solution();
        ListNode head = null;
        var rand = new Random();
        for (var i = 0; i < 20; i++)
        {
            head = new ListNode(rand.Next(20)) { next = head };
        }

        Console.WriteLine(head);
        Console.WriteLine(s.SortList(head));
    }
}

//  https://leetcode.com/submissions/detail/61085899/
//
//  Submission Details
//  15 / 15 test cases passed.
//      Status: Accepted
//      Runtime: 204 ms
//          
//          Submitted: 0 minutes ago
//
public class Solution {
    public ListNode SortList(ListNode head) {
        if (head == null || head.next == null)
        {
            return head;
        }
        
        var slow = head;
        var fast = slow.next;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        var upper = slow.next;
        slow.next = null;
        return Merge(SortList(head), SortList(upper));
    }
    
    public ListNode Merge(ListNode a, ListNode b)
    {
        var resultHead = new ListNode(-1);
        var result = resultHead;
        while (a != null || b != null)
        {
            if (a != null && (b == null || a.val < b.val))
            {
                result.next = a;
                a = a.next;
            }
            else
            {
                result.next = b;
                b = b.next;
            }
            
            result = result.next;
        }
        
        return resultHead.next;
    }
}

File: LeetCode/remote/sort_list.rb
#   https://leetcode.com/problems/sort-list/
#
#   Sort a linked list in O(n log n) time using constant space complexity.

#   
#   Submission Details
#   15 / 15 test cases passed.
#       Status: Accepted
#       Runtime: 192 ms
#           
#           Submitted: 0 minutes ago
#
#   https://leetcode.com/submissions/detail/53251589/

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def sort_list(head)
    return head if head.nil? || head.next.nil?
    pivot = head.val
    smaller = ListNode.new(-1)
    larger = ListNode.new(-1)
    equal = ListNode.new(-1)
    smaller_end = smaller
    larger_end = larger
    equal_end = equal
    while !head.nil?
        if head.val == pivot
            equal_end.next = head;
            equal_end = equal_end.next
        elsif head.val < pivot
            smaller_end.next = head
            smaller_end = smaller_end.next
        else
            larger_end.next = head
            larger_end = larger_end.next
        end
        
        head = head.next
    end
    
    smaller_end.next = nil
    equal_end.next = nil
    larger_end.next = nil
    
    smaller.next = sort_list(smaller.next)
    larger.next = sort_list(larger.next)
    
    it = smaller
    while !it.next.nil?
        it = it.next
    end
    
    it.next = equal.next;
    while !it.next.nil?
        it = it.next
    end
    
    it.next = larger.next
    return smaller.next
end

#   https://leetcode.com/submissions/detail/55374749/
#
#   Submission Details
#   15 / 15 test cases passed.
#       Status: Accepted
#       Runtime: 320 ms
#           
#           Submitted: 0 minutes ago
def sort_list(head)
    return head if head.nil? || head.next.nil?
    half = head
    finish = head.next
    while finish && finish.next
        half = half.next
        finish = finish.next.next
    end
    
    mid = half.next
    half.next = nil
    return merge(sort_list(head), sort_list(mid))
end

def merge(a, b)
    return nil if a.nil? && b.nil?
    return a if b.nil?
    return b if a.nil?
    if a.val < b.val
        a.next = merge(a.next, b)
        return a
    end
    
    b.next = merge(a, b.next)
    return b
end
